squashfs-tools/unsquashfs.c

```diff
@@ -2289,27 +2289,19 @@ int read_super(char *source)
 	 * Try to read a Squashfs 4 superblock
 	 */
 	int res = read_super_4(&s_ops);
-
 	if(res != -1)
 		return res;
 	else {
-		res = read_super_3(source, &s_ops, &sBlk_3);
-
-		if(res != -1)
-			return res;
-		else  {
-			res = read_super_2(&s_ops, &sBlk_3);
-
-			if(res != -1)
-				return res;
-
-			else {
-				res = read_super_1(&s_ops, &sBlk_3);
-
-				if(res != -1)
-					return res;
-			}
-		}
+        res = read_super_3(source, &s_ops, &sBlk_3);
```

Comment:
should we also check res? What if there is an other read error?

Reply:
Good catch. I changed the structure to make it easier to reason about and added a check for `res`.